Completed DCGAN tutorial

margaretmz opened the pull request:
I have completed the DCGAN tutorial with detailed explanations, a diagram and learning resources. Please review and approve the pull request.

margaretmz committed:
Updates to DCGAN  tutorial:

- Added TOC
- Added intro paragraph
- Added explanations of code
- Added learn more about GANs section
- Added a GANs network architecture diagram

alextp commented:
I somehow can't get github's notebook visualizer to show me this file. It seems to be a single line of code as opposed to the many lines before your edit; do you know if there's a way to reexport the ipynb?

margaretmz committed:
Reexport notebook to better reflect diff

margaretmz commented:
@alextp, I was able to show the notebook through the github notebook visualizer although at times it doesn't seem to work. I have reexported the notebook to better reflect the code diff. Looking forward to your review and feedback.

alextp commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,78 +1,186 @@
+      },
+      "cell_type": "markdown",
+      "source": [
+        "This tutorial demonstrates how to generate images of handwritten digits with **Deep Convolutional Generative Adverserial Networks** ([DCGAN](https://arxiv.org/pdf/1511.06434.pdf)). The code is written in [tf.keras](https://www.tensorflow.org/programmers_guide/keras) with [eager execution](https://www.tensorflow.org/programmers_guide/eager) enabled. "
```
adverserial -> adversarial

alextp commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,78 +1,186 @@
       "source": [
-        "This notebook demonstrates how to generate images of handwritten digits using [tf.keras](https://www.tensorflow.org/programmers_guide/keras) and [eager execution](https://www.tensorflow.org/programmers_guide/eager). To do so, we use Deep Convolutional Generative Adverserial Networks ([DCGAN](https://arxiv.org/pdf/1511.06434.pdf)).\n",
+        "## What are GANs?\n",
+        "GANs standards for **Generative Adversarial Networks** and they are a type of deep **generative** models. MIT [Intro to Deep Learning](http://introtodeeplearning.com/) lecture on **Deep Generative Models** has a great intro to generative models as well as GANs. ([video](https://youtu.be/JVb54xhEw6Y) | [slides](http://introtodeeplearning.com/materials/2018_6S191_Lecture4.pdf)). We have in ...
```
"standards for" -> "stands for"

alextp commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,78 +1,186 @@
+        " **Deep Convolutional GAN** (DCGAN) is a type of GANs and in this tutorial we will use DCGAN to generate MNIST digits.\n",
         "\n",
-        "Below is the output generated after training the generator and discriminator models for 150 epochs.\n",
+        "GANs can be used to generate new images that no one has seen before. The generator will generate fake images while the discriminator will classify whether the generated images are fake. An** equilibrium** will reach in the game when the generator makes data that looks identical to the training data and the discriminator can no longer tell the difference between the fake images (generated ...
```
"an equilibrium will reach in the game" -> "the game will reach an equilibrium"

alextp commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,78 +1,186 @@
+        "\n",
+        "![alt text](https://github.com/margaretmz/tensorflow/blob/margaret-dcgan/tensorflow/contrib/eager/python/examples/generative_examples/gans_diagram.png?raw=1)\n",
+        "\n",
+        "While the generator and discriminator competes against each other, the discriminator also teaches the generator . Over time the generator starts to produce images that resemble the training data that is fed into the discriminator, in this case the MNIST digits. Below is the output with images generated after training the generator and discriminator models for 150 epochs.\n",
```
"competes against" -> "compete against"

alextp commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -327,96 +517,100 @@
+      },
       "cell_type": "markdown",
+      "source": [
+        "Now it's time to put together the generator and discriminator to set up the Generative Adversarial Networks, as you wee the diagam in the beginning of the tutorial."
```
I think you did not mean to say "as you wee the diagram". See?

alextp reviewed (changes requested):
Looks much improved; thanks! I only have a few minor spelling nits.

margaretmz committed:
Fix typos and wording based on review feedback

margaretmz commented:
Thanks @alextp for your review and feedback! I made updates accordingly.

asimshankar commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -84,114 +192,135 @@
-      "outputs": [],
+      "cell_type": "code",
       "source": [
         "class Generator(tf.keras.Model):\n",
```
Both this and the discriminator class seems like they will be well served by `tf.keras.Sequential`, making the code even more compact. Should we write them instead as:

```python
def make_generator_model():
  return tf.keras.Sequential([
    tf.keras.layers.Dense(7*7*64, use_bias=False),
    tf.keras.layers.BatchNormalization(),
    tf.keras.layers.Relu(),
    ...])
```

random-forests committed:
Updates to the text.

random-forests commented:
Thank you @margaretmz! I added some edits to the text (it was easier to commit the source of my Jupyter Notebook rather than copy / pasting them in, I hope I didn't clobber anything!)

margaretmz commented:
Thank you @asimshankar for suggesting we use the Sequential model instead of model subclassing for the Generator and Discriminator classes. The current code is using  tf.contrib.eager.defun to boost performance. Can we still use defun with the Sequential model? My understanding is that defun is typically applied in the forward computation, backward computation for the gradients and applications of gradients to variables.

margaretmz commented:
Thanks @random-forests for your edits to make the tutorial more concise! I verified that the edited notebook looks good.

random-forests committed:
Adding note on model subclassing vs sequential

alextp commented:
@margaretmz defun and sequential should work well together, so this should be fine.

random-forests commented:
@asimshankar, thanks for the suggestion on sequential. I added a note about that in the notebook. In this case, I'd prefer to keep subclassing (even though its more verbose), just so we have more examples on tf.org in this style - in case it'll help folks modify this for other projects down the road.

alextp commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,694 +1,928 @@
+   },
+   "outputs": [],
+   "source": [
+    "from __future__ import absolute_import, division, print_function\n",
```
Can we ditch the future imports? I think they're sad.

margaretmz replied:
Removed the future imports.

alextp commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,694 +1,928 @@
+    "      discriminator_optimizer.apply_gradients(zip(gradients_of_discriminator, discriminator.variables))\n",
+    "\n",
+    "      \n",
+    "    if epoch % 1 == 0:\n",
```
epoch % 1 == 0 is a tautology

margaretmz replied:
Removed.

random-forests commented:
Silly CLA bot.

asimshankar commented:
@margaretmz : Yup, it should work, see https://github.com/tensorflow/tensorflow/commit/96c7397e08a8fde1bdf069b4e7dd74b5a8d79a31 :)

@random-forests : I have a slight preference for using `Sequential` here since I think we do want to encourage that style, not only because of brevity, but also because it provides a bunch of other benefits in terms of serialization, introspection (e.g., iterating over the layer structure of a network) etc. Till a couple of months ago there were a bunch of pain points with `Sequential` in eager that have since been addressed. Long story short: I think using `Sequential` here will be good for users and instructive. But if you still feel strongly, I won't object.

margaretmz committed:
Updates for alextp feedback

margaretmz commented:
I've made updates for your requested changes, @alextp.

Thanks @alextp and @asimshankar for confirming that Sequential and defun work well together. I agree with @random-forests - what we have right now in the tutorial will let the readers see both styles and there are plenty of examples on Sequential already.

margaretmz committed:
Reset file to before Sequential update

margaretmz committed:
Update GANs models to use Sequential and 50 epochs

margaretmz commented:
@asimshankar, @alextp and @random-forests - I updated the generator and discriminator models to use Sequential, and updated the text accordingly. I modified the generator model during the conversion and the network now generates resembling digits after only 50 epochs. Please review and provide your feedback. Thanks.

asimshankar commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,324 +1,428 @@
+        "    tf.keras.layers.LeakyReLU(),\n",
+        "      \n",
+        "    tf.keras.layers.Reshape((7, 7, 256)),\n",
+        "    # Reshape layer to 7x7x64 \n",
```
Optional: Instead of these comments, I wonder if it would make sense to use the Keras model's `output_shape` property and assertions.  That way the comment and code are always in sync. Though, you'll need to provide the `input_shape` argument to the first layer.

Something like:

```python
def make_generator_model():
  m = tf.keras.Sequential()
  m.add(tf.keras.layers.Dense(7*7*256, use_bias=False, input_shape=(100,))
  m.add(tf.keras.layers.BatchNormalization())
  m.add(tf.keras.layers.LeakyReLU())
  m.add(tf.keras.layers.Conv2DTranspose(128, (5, 5), strides=(1, 1), padding='same', use_bias=False))
  assert m.output_shape == (None, 7, 7, 64)
  m.add(...)
  ...
  assert m.output_shape == (None, 28, 28, 1)
  return m
```
(Actually, in this process, I noticed that the comments are a bit off :))

margaretmz replied:
@asimshankar, thanks for your suggestion! I made changes as you suggested.
It's indeed better than comments - yesterday I forgot to update the comment after I change the code.

asimshankar commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,324 +1,428 @@
-        "generator = Generator()\n",
-        "discriminator = Discriminator()"
+        "\n",
+        "This model takes about ~30 seconds per epoch to train on a single Tesla K80 on Colab, as of July 2018. Eager execution can sometimes be slower than executing the equivalent graph due to overheads of interpreting Python code. By using [tf.contrib.eager.defun](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun) to create graph functions, we get a ~10 secs/epoch performance b ...
```
Optional: Instead of `defun`ing the model calls, did you try `defun`-ing the whole training step? That might yield some good improvements too (and will be simpler)?

So instead of:

```python
def train(dataset, epochs, noise_dim):
  for epoch in range(epochs):
    for images in dataset:
      ...
```

You could have:

```python
def train_step(images):
   noise = tf.random_normal(...)
   with tf.GradientTape() as tape:
      ...
   

def train(dataset, epochs):
  for epoch in epochs:
    for images in dataset:
      train_step(images)
    # All that display stuff
```

Then for faster training you could do:
```python
train_step = tf.contrib.eager.defun(train_step)
```

margaretmz replied:
Updated to defun on train_steps instead of model calls. Performance boost is better at ~20 secs / epoch instead of ~10 secs / epoch (defun on model calls). Thanks!

asimshankar commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -327,142 +431,134 @@
+        "colab": {}
       },
-      "outputs": [],
+      "cell_type": "code",
```
Why pass `noise_dim` as an argument to `train()` given that it is defined as almost a constant above?
The module level `noise_dim` constant and the function argument can be confusing. How about just having a `NOISE_DIM` constant and using that instead of creating a constant and then providing that as an argument of the same name to `train()`?

margaretmz replied:
Good catch. Fixed.

margaretmz committed:
Add assert for generator layer output shape and defun train_sstep instead of model calls

margaretmz committed:
Remove out of place #

asimshankar commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,324 +1,405 @@
+        "    model.add(tf.keras.layers.BatchNormalization())\n",
+        "    model.add(tf.keras.layers.LeakyReLU())\n",
+        "      \n",
+        "    model.add(tf.keras.layers.Reshape((7, 7, 256)))\n",
```
Optional: Maybe a comment here mentioning that `None` is the batch dimension?

margaretmz replied:
Done

asimshankar commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -1,324 +1,405 @@
       "cell_type": "code",
+      "source": [
+        "def make_discriminator_model():\n",
+        "    model = tf.keras.Sequential()\n",
```
Optional: Since there are no asserts here, the original version (where the list was provided to the constructor) works just as well. Your call either way.

margaretmz replied:
I left it as is because it's easier for the reader to experiment with assert, print or model.summary() etc. if they want to.

asimshankar reviewed (dismissed):
Looks good to me, thanks!

asimshankar commented on tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb:
```diff
@@ -327,170 +408,200 @@
-        "      generate_and_save_images(generator,\n",
+        "This model takes about ~30 seconds per epoch to train on a single Tesla K80 on Colab, as of October 2018. \n",
+        "\n",
+        "Eager execution can sometimes be slower than executing the equivalent graph due to overheads of interpreting Python code. By using [tf.contrib.eager.defun](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun) to create graph functions, we get a ~20 secs/epoch performance boost (from ~50 secs/epoch down to ~30 secs/epoch). This way we get the best of both eager execution (ea ...
```
I'd rephrase this a bit (so as to not blame just the Python interpreter). There may be other graph optimizations we're benefiting from here as well (constant folding, common subexpression elimination, layout changes etc.). Though, in this particular case it is probably the Python overhead (since we haven't enabled grappler passes on function yet). How about:

"Eager execution can be slower than executing the equivalent graph as it can't benefit from whole-program optimizations on the graph, and also incurs overheads of interpreting Python code. By using ..."

margaretmz replied:
Updated as suggested.

margaretmz committed:
Added comment that None is the batch size

margaretmz committed:
Updated text on eager execution

margaretmz committed:
Fix text regarding input noise is sampled from a normal distribution

margaretmz commented:
@asimshankar, @alextp and @random-forests, I have made updates per all your review feedback. Looking forward to your approval and PR merge. Thanks.
